kotlin/tracing-debugger/src/main/java/com/squareup/workflow/tracing/TraceEvent.kt

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow.tracing
+
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.ASYNC_BEGIN
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.ASYNC_END
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.DURATION_BEGIN
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.DURATION_END
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.INSTANT
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.OBJECT_CREATED
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.OBJECT_DESTROYED
+import com.squareup.workflow.tracing.ChromeTraceEvent.Phase.OBJECT_SNAPSHOT
+import com.squareup.workflow.tracing.ChromeTraceEvent.Scope.PROCESS
+import com.squareup.workflow.tracing.TraceEvent.AsyncDurationBegin
+import com.squareup.workflow.tracing.TraceEvent.AsyncDurationEnd
+import com.squareup.workflow.tracing.TraceEvent.DurationBegin
+import com.squareup.workflow.tracing.TraceEvent.DurationEnd
+import com.squareup.workflow.tracing.TraceEvent.Instant
+import com.squareup.workflow.tracing.TraceEvent.ObjectCreated
+import com.squareup.workflow.tracing.TraceEvent.ObjectDestroyed
+import com.squareup.workflow.tracing.TraceEvent.ObjectSnapshot
+
+/**
+ * Represents a single event in a trace. Write a stream of lists of these to a sink with
+ * [streamTraceTo].
+ */
+sealed class TraceEvent {
```

Comment:
Played with the idea of making this API much more stateful to enforce things like "all duration events must be correctly nested, or they'll show up weird in the UI" but backed away because it just adds a bunch of overhead that's not actually necessary when the events are already guaranteed to be delivered in the correct order, and I'm trying to avoid as much overhead as possible here.